### PR TITLE
[FEA] hook-based support for distributed but shared parameter

### LINKED ISSUE #235: 🚀[FEA]: Add a wrapper class for shared tensors in model parallel implementations
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Critical (currently preventing usage)

### Please provide a clear description of problem you would like to solve.

https://github.com/NVIDIA/modulus/pull/153 recommends using DDP with a `"data_parallel"` process group that is orthogonal to the `"model_parallel"` process group across which the weights of a model are distributed. This will perform the gradient all-reduce only along the `"data_parallel"` process group which is the right behavior for distributed weights.

However, in many cases, even with model parallelism, some weights of a model are shared across the `"model_parallel"` group. For these cases, the gradient reduction should also happen along the `"model_parallel"` process group. This can be done in two ways:
1. Registering a custom [communication hook](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.register_comm_hook) that performs the additional reduction along the `"model_parallel"` group for parameters that are shared. See [SFNO implementation](https://github.com/NVIDIA/modulus/blob/main/modulus/utils/sfno/distributed/mappings.py#L240) for specifics.
2. Adding custom ops to the model implementation that are no-ops in the forward pass but do the gradient reduction across the shared process group in the backward pass.

In both cases, the model implementation somehow needs to indicate which tensors are shared and across what groups. It would be ideal if whatever this mechanism ends up being also automatically sets up gradient hooks or the backward reduction op. One proposed solution is to create a class wrapper that can also be used as a decorator. Something like
```python
m = SharedModel(nn.Linear(20, 30))
```
and use as usual
```python
output = m(input)
```

A rough sketch of the `SharedModel` wrapper class implementation:
```python
import modulus
from modulus.distributed import DistributedManager
from modulus.distributed.utils import _reduce
import torch
from torch import nn


class SharedModelBackwardReduction(torch.autograd.Function):

    @staticmethod
    def symbolic(graph, input_, group_):
        return input_

    @staticmethod
    def forward(ctx, input_, group_):
        ctx.group = group_
        return input_

    @staticmethod
    def backward(ctx, grad_output):
        return _reduce(grad_output, group=DistributedManager().group(ctx.group))


class SharedModel(nn.Module):
    def __init__(self, model: Union[nn.Module, modulus.Module], group=None):
        self.group = group
        self.model = model
     
    def forward(self, *args, **kwargs):
        # Set up forward no-op for backward reduction
        for param in self.model.parameters():
            SharedModelBackwardReduction.apply(param, self.group)

        return self.model(*args, **kwargs)
```

### Describe any alternatives you have considered

_No response_

### PR DESCRIPTION
**Pull request by stadlmax**
# Modulus Pull Request

## Description
#235  asks for support for distributed parameters which are shared across the model-parallel group. Initially, there was the idea of using a wrapper idea. The hook-based approach, however, should be less intrusive and more flexible and is introduced in this draft. Shared weights are simply marked or unmarked as needed which registers a gradient hook taking care of the necessary reduction of gradients.

closes #235 , but with a different implementation idea

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: test/distributed/test_utils.py [hunk 1]
```diff
@@ -14,23 +14,210 @@
...
+        ref_module.lin_1.weight.grad.clone().detach(),
+        group=manager.group("shared_parallel"),
+        use_fp32=True,
+    )
+    ref_lin_1_bias_grad = _reduce(
+        ref_module.lin_1.bias.grad.clone().detach(),
+        group=manager.group("shared_parallel"),
+        use_fp32=True,
+    )
+    ref_lin_2_weight_grad = _reduce(
```
Comment (akshaysubr) [inline test/distributed/test_utils.py:97]: Nit: Can lin_2 reductions be moved to below the first set of asserts to make the test a bit easier to follow?

File: test/distributed/test_utils.py [hunk 2]
```diff
@@ -14,23 +14,210 @@
...
+    dist_out = dist_module(x)
+    dist_out.backward(torch.ones_like(dist_out))
+    assert torch.allclose(ref_lin_2_weight_grad, dist_module.lin_2.weight.grad)
+    assert torch.allclose(ref_lin_2_bias_grad, dist_module.lin_2.bias.grad)
+    assert torch.allclose(ref_module.lin_1.weight.grad, dist_module.lin_1.weight.grad)
+    assert torch.allclose(ref_module.lin_1.bias.grad, dist_module.lin_1.bias.grad)
+
+    # unmark lin_2 again (unmarking lin_1 should throw an error)
+    try:
+        unmark_module_as_shared(dist_module.lin_1)
```
Comment (akshaysubr) [inline test/distributed/test_utils.py:148]: Same comment here about using `pytest.raises`

File: modulus/distributed/utils.py [hunk 3]
```diff
@@ -670,3 +671,58 @@ def indexed_all_to_all_v_wrapper_bwd(
...
+) -> nn.Module:
+    group = DistributedManager().group(process_group)
+    handle_key = "_shared_weight_dist_hook"
+
+    def hook(grad: torch.Tensor) -> torch.Tensor:
+        # the documentation states that
+        # "The hook should not modify its argument, but it can optionally return a new gradient
+        #  which will be used in place of grad."
+        # as all_reduce is an in-place operation, need to copy gradient
+        grad = _reduce(grad.clone().detach(), group=group, use_fp32=use_fp32_reduction)
```
[the remark this answers is not in the document]
Comment (stadlmax) [reply modulus/distributed/utils.py:690]: ah, actually good question, we need the copy, but the hook shouldn't be part of the compute graph meaning we don't have to take care of that.

File: modulus/distributed/utils.py [hunk 4]
```diff
@@ -670,3 +671,58 @@ def indexed_all_to_all_v_wrapper_bwd(
...
+        #  and not the grad itself. The hook can in-place modify and access its Tensor argument,
+        # including its .grad field."
+        param.grad = _reduce(param.grad, group=group, use_fp32=use_fp32_reduction)
+
+    for name, param in module.named_parameters(recurse=recurse):
+        error_msg = f"Parameter {name} already marked as having shared weights, can't mark it again!"
+        if hasattr(param, handle_key):
+            raise RuntimeError(error_msg)
+        if torch.__version__ < (2, 1):
+            handle = param.register_hook(hook)
```
[the remark this answers is not in the document]
Comment (stadlmax) [reply modulus/distributed/utils.py:705]: It was introduced in https://github.com/pytorch/pytorch/pull/107063
which to me seems like it is only available from 2.1 onwards, right?

File: test/distributed/test_utils.py [hunk 5]
```diff
@@ -14,23 +14,210 @@
...
     # Reset class state
     DistributedManager._shared_state = {}
     DistributedManager.initialize()
 
     manager = DistributedManager()
     assert manager.is_initialized()
 
-    loss = gather_loss(1.0, dst_rank=0, mean=True)
-    assert loss == 1.0
+    loss = gather_loss(1.0, dst_rank=0, mean=False)
```
[the remark this answers is not in the document]
Comment (stadlmax) [reply test/distributed/test_utils.py:43]: good point, i actually didn't touch this per se, I just included the test again

[the remark this answers is not in the document]
Comment (stadlmax) [reply test/distributed/test_utils.py:43]: I changed this to `reduce_loss` now.

File: modulus/distributed/utils.py [hunk 6]
```diff
@@ -670,3 +671,58 @@ def indexed_all_to_all_v_wrapper_bwd(
         out = out.to(tensor.dtype)
 
     return out
+
+
+def mark_module_as_shared(
+    module: nn.Module,
+    process_group: Optional[str],
```
[the remark this answers is not in the document]
Comment (akshaysubr) [reply modulus/distributed/utils.py:670]: That's a good point. Can leave it as is then.

**Review by akshaysubr — APPROVED** [review]

**Review by mnabian — APPROVED** [review]
